drivers/adc/adc_sam0.c

```diff
@@ -579,11 +625,9 @@ static const struct adc_driver_api adc_sam0_api = {
 	static void adc_sam0_config_##n(struct device *dev);		      \
 	static const struct adc_sam0_cfg adc_sam_cfg_##n = {		      \
 		.regs = (Adc *)DT_ATMEL_SAM0_ADC_ADC_##n##_BASE_ADDRESS,      \
+		.clk_dev = DT_ATMEL_SAM0_ADC_ADC_##n##_CLOCK_CONTROLLER,      \
```

Comment:
IMHO information about clock driving given peripheral should be managed by OS, not the driver.
Ideally, we should have something like device_clock_get(dev, idx).

Reply:
I agree that having that kind of API would be very useful, I'm not sure how you'd actually implement it.  You'd somehow need to attach a dynamic number of references to the `struct device` while also avoiding a combinatorial explosion with the device definitions (e.g. you already have to deal with the effects of `CONFIG_DEVICE_POWER_MANAGEMENT` on device declaration, even adding a single clock would take the permutations to four).

On a higher level, it seems like a more general "get associated device" API would be useful.  There's at least two cases of "sibling" devices currently handled like what I'm doing here (other `clock_control` usage and DMA).  I could also see an argument for having the same API get the parent and children devices on a bus.  I have no idea how you'd implement that, though.

In either case, this (and the other cases that I copied the approach from) could be easily adapted to that kind of API if it became available.